### PR TITLE
5840 OBX Segment from FHIR to HL7

### PR DESCRIPTION
**Pull request by TurkeyFried**
This PR generates the HL7 OBX segment from FHIR data.

Test Steps:
1. Check out the branch for this PR.
2. Run the following command to convert a FHIR bundle to HL7.  
`./prime fhirdata --input-file "src/testIntegration/resources/datatests/HL7_to_FHIR/sample_co_1_20220518-0001.fhir" -s metadata/hl7_mapping/ORU_R01/ORU_R01-base.yml`
3. Verify an HL7 message is generated from the conversion and includes the OBX segments.
4. Look at the source HL7 used at `src/testIntegration/resources/datatests/HL7_to_FHIR/sample_co_1_20220518-0001.hl7` and compare the data for the OBX segments.
5. Verify both OBX segments are identical, except for any caveats in the comparison spreadsheet on Sharepoint.

## Changes
- Added config to handle OBX conversion from the FHIR data.
- Made a few updates on the HL7 to FHIR conversion to properly store some data.

### Testing
- [x] Tested locally?
- [ ] Ran `./prime test` or `./gradlew testSmoke` against local Docker ReportStream container?
- [ ] (For Changes to /frontend-react/...) Ran `npm run lint:write`? 
- [ ] Added tests?

### Process
- [ ] Are there licensing issues with any new dependencies introduced?
- [ ] Includes a summary of what a code reviewer should test/verify?
- [ ] Updated the release notes?
- [ ] Database changes are submitted as a separate PR?
- [ ] DevOps team has been notified if PR requires ops support?

## Linked Issues
- #5840

## To Be Done
*Create GitHub issues to track the work remaining, if any*

## Specific Security-related subjects a reviewer should pay specific attention to
- Does this PR introduce new endpoints?
    - new endpoint A
    - new endpoint B
- Does this PR include changes in authentication and/or authorization of existing endpoints?
- Does this change introduce new dependencies that need vetting?
- Does this change require changes to our infrastructure?
- Does logging contain sensitive data?
- Does this PR include or remove any sensitive information itself?

If you answered '_yes_' to any of the questions above, conduct a detailed Review that addresses at least:

- What are the potential security threats and mitigations? Please list the _STRIDE_ threats and how they are mitigated
    - **S**poofing (faking authenticity)
        - Threat _T_, which could be achieved by _A_, is mitigated by _M_
    - **T**ampering (influence or sabotage the integrity of information, data, or system)
    - **R**epudiation (the ability to dispute the origin or originator of an action)
    - **I**nformation disclosure (data made available to entities who should not have it)
    - **D**enial of service (make a resource unavailable)
    - **E**levation of Privilege (reduce restrictions that apply or gain privileges one should not have)
- Have you ensured logging does not contain sensitive data?
- Have you received any additional approvals needed for this change?

### CODE REVIEW

File: prime-router/metadata/fhir_mapping/hl7/codesystem/ExtensionUrlMapping.yml [hunk 1]
```diff
@@ -190,4 +190,7 @@
   url: "https://reportstream.cdc.gov/fhir/StructureDefinition/placer-assigned-identifier"
 # Filler assigned identifier
 - id: "filler-assigned-identifier"
-  url: "https://reportstream.cdc.gov/fhir/StructureDefinition/filler-assigned-identifier"
\ No newline at end of file
+  url: "https://reportstream.cdc.gov/fhir/StructureDefinition/filler-assigned-identifier"
+# Producer's ID
```
Comment (carlosfelix2) [inline prime-router/metadata/fhir_mapping/hl7/codesystem/ExtensionUrlMapping.yml:194]: Change this comment to just sub ID.  It is generic enough to be use in multiple places

File: prime-router/metadata/hl7_mapping/ORU_R01/base/observation-result.yml [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+  - name: result-set-ID
+    value: [ '%resultIndex + 1' ]
+    hl7Spec: [ '%{hl7OBXField}-1' ]
+
+  - name: result-value-type-st
+    value: [ '''ST''' ]
+    hl7Spec: [ '%{hl7OBXField}-2' ]
+
+  - name: result-value-type-cwe
+    condition: '%resource.value.exists()'
```
Comment (carlosfelix2) [inline prime-router/metadata/hl7_mapping/ORU_R01/base/observation-result.yml:14]: Change this to `%resource.value is CodeableConcept`

File: prime-router/metadata/hl7_mapping/ORU_R01/base/observation-result.yml [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    hl7Spec: [ '%{hl7OBXField}-3-2' ]
+
+  - name: observation-identifier-coding-system
+    value: [ '%resource.code.coding.version' ]
+    hl7Spec: [ '%{hl7OBXField}-3-3' ]
+
+  # override to handle LN translated to loinc L on HL7->FHIR conversion
+  - name: observation-identifier-coding-system-ln
+    resource: '%resource.code.coding'
+    condition: '%resource.system = "http://loinc.org" and %resource.version = "L"'
```
Comment (carlosfelix2) [inline prime-router/metadata/hl7_mapping/ORU_R01/base/observation-result.yml:33]: I think Patricia added something for this.  Check out her code.

File: prime-router/metadata/hl7_mapping/ORU_R01/base/observation-result.yml [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    value: [ '%resource.extension(%`rsext-sub-id`).value' ]
+    hl7Spec: [ '%{hl7OBXField}-4' ]
+
+  - name: observation-value-st
+    condition: '%resource.value is string'
+    value: [ '%resource.value' ]
+    hl7Spec: [ '%{hl7OBXField}-5' ]
+
+  - name: observation-value-cwe-identifier
+    condition: '%resource.value.coding.exists()'
```
Comment (carlosfelix2) [inline prime-router/metadata/hl7_mapping/ORU_R01/base/observation-result.yml:59]: Change this to `%resource.value is CodeableConcept`

File: prime-router/metadata/hl7_mapping/ORU_R01/base/observation-result.yml [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    value: [ '%resource.value.coding.code' ]
+    hl7Spec: [ '%{hl7OBXField}-5-1' ]
+
+  - name: observation-value-cwe-identifier
+    condition: '%resource.value.exists()'
```
Comment (carlosfelix2) [inline prime-router/metadata/hl7_mapping/ORU_R01/base/observation-result.yml:64]: Change to `%resource.value is CodeableConcept`

**Review by carlosfelix2 — CHANGES_REQUESTED** [review]
OBX 2 and higher have changed or missing data.  Use the string comparison tool to see it.  https://text-compare.com/

File: prime-router/metadata/hl7_mapping/ORU_R01/base/observation-result.yml (hunk 2 above)
Comment (TurkeyFried) [reply prime-router/metadata/hl7_mapping/ORU_R01/base/observation-result.yml]: fixed

File: prime-router/metadata/hl7_mapping/ORU_R01/base/observation-result.yml (hunk 4 above)
Comment (TurkeyFried) [reply prime-router/metadata/hl7_mapping/ORU_R01/base/observation-result.yml]: fixed

File: prime-router/metadata/hl7_mapping/ORU_R01/base/observation-result.yml (hunk 5 above)
Comment (TurkeyFried) [reply prime-router/metadata/hl7_mapping/ORU_R01/base/observation-result.yml]: fixed

File: prime-router/metadata/fhir_mapping/hl7/codesystem/ExtensionUrlMapping.yml (hunk 1 above)
Comment (TurkeyFried) [reply prime-router/metadata/fhir_mapping/hl7/codesystem/ExtensionUrlMapping.yml]: fixed

File: prime-router/metadata/hl7_mapping/ORU_R01/base/observation-result.yml (hunk 3 above)
Comment (TurkeyFried) [reply prime-router/metadata/hl7_mapping/ORU_R01/base/observation-result.yml:38]: I can't use it here. The mapping is just different enough that it doesn't play nice with it.

File: prime-router/metadata/hl7_mapping/ORU_R01/base/observation-result.yml [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+
+  - name: observation-identifier-code
+    value: [ '%resource.code.coding.code' ]
+    hl7Spec: [ '%{hl7OBXField}-3-1' ]
+
+  - name: observation-identifier-text
+    value: [ '%resource.code.text' ]
+    hl7Spec: [ '%{hl7OBXField}-3-2' ]
+
+#  - name: observation-identifier-coding-system
```
Comment (carlosfelix2) [inline prime-router/metadata/hl7_mapping/ORU_R01/base/observation-result.yml:27]: Remove any commented out elements

File: prime-router/metadata/hl7_mapping/ORU_R01/base/observation-result.yml [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+    resource: '%resource.performer.resolve().practitioner.resolve()'
+    constants:
+      hl7XCNField: '%{hl7OBXField}-25'
+    schema: xcn-contact
+    resourceIndex: contactIndex
+
+  - name: observation-note
+    resource: '%resource.note.text.split(''\n'')'
+#    condition: >-
+#      Bundle.entry.resource.ofType(DiagnosticReport)[%orderIndex].result[%observationIndex].resolve().note.exists()
```
Comment (carlosfelix2) [inline prime-router/metadata/hl7_mapping/ORU_R01/base/observation-result.yml:136]: Remove any commented out elements

File: prime-router/src/main/kotlin/fhirengine/translation/hl7/utils/ConstantResolver.kt [hunk 8]
```diff
@@ -270,6 +271,25 @@ object CustomFHIRFunctions {
         }
     }
 
+    /**
```
Comment (carlosfelix2) [inline prime-router/src/main/kotlin/fhirengine/translation/hl7/utils/ConstantResolver.kt:274]: Remove all of this stuff from your PR.  Make sure to go over all the changes in the PR, so you can identify anything that looks weird

**Review by carlosfelix2 — CHANGES_REQUESTED** [review]

File: prime-router/metadata/hl7_mapping/ORU_R01/base/observation-result.yml (hunk 7 above)
Comment (TurkeyFried) [reply prime-router/metadata/hl7_mapping/ORU_R01/base/observation-result.yml]: oops, removed

File: prime-router/metadata/hl7_mapping/ORU_R01/base/observation-result.yml (hunk 6 above)
Comment (TurkeyFried) [reply prime-router/metadata/hl7_mapping/ORU_R01/base/observation-result.yml]: fixed

File: prime-router/src/main/kotlin/fhirengine/translation/hl7/utils/ConstantResolver.kt (hunk 8 above)
Comment (TurkeyFried) [reply prime-router/src/main/kotlin/fhirengine/translation/hl7/utils/ConstantResolver.kt]: done

**Pull request merged**